File: examples/external/lua-clock/widget.lua
-- lua-clock -- xtop external widget.
--
-- Speaks the xtop line-delimited JSON protocol on stdin/stdout: one JSON
-- request object per line in, one JSON response object per line out.
-- Standard Lua only; the codec lives in json.lua next to this file.

-- Resolve json.lua relative to this script so any cwd works.
local script = arg and arg[0] or "widget.lua"
local script_dir = script:match("^(.*[/\\])") or ""
if script_dir ~= "" then
  package.path = script_dir .. "?.lua;" .. package.path
end
local json = require("json")

local WIDGET_NAME = "lua-clock"
local VERSION = "0.1.0"

local COLOR_TITLE = { 200, 200, 200 }
local COLOR_CLOCK = { 123, 216, 143 }
local COLOR_DIM = { 140, 140, 140 }

local function send(obj)
  io.stdout:write(json.encode(obj), "\n")
  io.stdout:flush()
end

local function send_log(message)
  send({ type = "log", message = tostring(message) })
end

-- "3d 04:05:06" or "04:05:06" from uptime seconds.
local function format_uptime(seconds)
  seconds = tonumber(seconds) or 0
  local days = math.floor(seconds / 86400)
  local rest = seconds % 86400
  local hours = math.floor(rest / 3600)
  local minutes = math.floor((rest % 3600) / 60)
  local secs = rest % 60
  if days > 0 then
    return string.format("%dd %02d:%02d:%02d", days, hours, minutes, secs)
  end
  return string.format("%02d:%02d:%02d", hours, minutes, secs)
end

local function build_ops(state)
  local width = tonumber(state and state.width) or 0
  local height = tonumber(state and state.height) or 0
  if width <= 0 then
    width = 30
  end
  if height <= 0 then
    height = 6
  end

  local snapshot = (state and state.snapshot) or {}
  local unix_time = tonumber(state and state.unix_time) or 0
  local clock = os.date("!%H:%M:%S", unix_time) -- '!' = UTC
  local uptime = format_uptime(snapshot.uptime)
  local inner_width = math.max(width - 2, 1)
  local bottom = math.max(height - 2, 1)

  return {
    {
      op = "block",
      rect = { x = 0, y = 0, width = width, height = height },
      border = "rounded",
      title = WIDGET_NAME,
      fg = COLOR_TITLE,
      bg = json.null,
    },
    {
      op = "text",
      rect = { x = 1, y = 1, width = inner_width, height = 1 },
      spans = {
        { text = clock, fg = COLOR_CLOCK, bold = true },
      },
      align = "center",
      wrap = false,
    },
    {
      op = "text",
      rect = { x = 1, y = bottom, width = inner_width, height = 1 },
      spans = {
        { text = "uptime " .. uptime, fg = COLOR_DIM, dim = true },
      },
      align = "center",
      wrap = false,
    },
  }
end

-- Returns false when the guest should exit.
local function handle_request(request)
  local kind = request.type
  if kind == "manifest" then
    send({
      type = "manifest",
      manifest = {
        name = WIDGET_NAME,
        version = VERSION,
        description = "UTC clock and uptime rendered from the xtop state (Lua)",
        author = "xtop examples",
        max_processes = 1, -- this widget does not read processes
        api = "1",
      },
    })
  elseif kind == "render" then
    local ok, ops = pcall(build_ops, request.state)
    if not ok then
      send_log("render failed: " .. tostring(ops))
      send({ type = "draw", ops = {} })
    else
      send({ type = "draw", ops = ops })
    end
  elseif kind == "shutdown" then
    return false
  else
    send_log("unknown request type: " .. tostring(kind))
  end
  return true
end

-- EOF on stdin ends the loop cleanly.
for line in io.lines() do
  if line:match("%S") then
    local ok, request = pcall(json.decode, line)
    if not ok or type(request) ~= "table" then
      send_log("invalid JSON request: " .. tostring(request))
    elseif handle_request(request) == false then
      break
    end
  end
end

os.exit(0)

File: examples/external/node-clock/widget.js
#!/usr/bin/env node
// node-clock -- xtop external widget.
//
// Speaks the xtop line-delimited JSON protocol on stdin/stdout: one JSON
// request object per line in, one JSON response object per line out.
// Node.js standard library only.

'use strict';

const readline = require('readline');

const WIDGET_NAME = 'node-clock';
const VERSION = '0.1.0';

const COLOR_TITLE = [200, 200, 200];
const COLOR_CLOCK = [123, 216, 143];
const COLOR_DIM = [140, 140, 140];

function send(obj) {
  process.stdout.write(JSON.stringify(obj) + '\n');
}

function sendLog(message) {
  send({ type: 'log', message: String(message) });
}

// "3d 04:05:06" or "04:05:06" from uptime seconds.
function formatUptime(seconds) {
  const total = Math.max(0, Math.floor(Number(seconds) || 0));
  const days = Math.floor(total / 86400);
  const rest = total % 86400;
  const pad = (n) => String(n).padStart(2, '0');
  const hms = `${pad(Math.floor(rest / 3600))}:${pad(Math.floor((rest % 3600) / 60))}:${pad(rest % 60)}`;
  return days > 0 ? `${days}d ${hms}` : hms;
}

function utcClock(unixTime) {
  return new Date((Number(unixTime) || 0) * 1000).toISOString().slice(11, 19);
}

function buildOps(state) {
  if (state === null || typeof state !== 'object') state = {};
  const snapshot =
    state.snapshot !== null && typeof state.snapshot === 'object' ? state.snapshot : {};

  const width = Number(state.width) > 0 ? Number(state.width) : 30;
  const height = Number(state.height) > 0 ? Number(state.height) : 6;
  const clock = utcClock(state.unix_time);
  const uptime = formatUptime(snapshot.uptime);
  const innerWidth = Math.max(width - 2, 1);
  const bottom = Math.max(height - 2, 1);

  return [
    {
      op: 'block',
      rect: { x: 0, y: 0, width, height },
      border: 'rounded',
      title: WIDGET_NAME,
      fg: COLOR_TITLE,
      bg: null,
    },
    {
      op: 'text',
      rect: { x: 1, y: 1, width: innerWidth, height: 1 },
      spans: [{ text: clock, fg: COLOR_CLOCK, bold: true }],
      align: 'center',
      wrap: false,
    },
    {
      op: 'text',
      rect: { x: 1, y: bottom, width: innerWidth, height: 1 },
      spans: [{ text: 'uptime ' + uptime, fg: COLOR_DIM, dim: true }],
      align: 'center',
      wrap: false,
    },
  ];
}

// Returns false when the guest should exit.
function handle(request) {
  const kind = request.type;
  if (kind === 'manifest') {
    send({
      type: 'manifest',
      manifest: {
        name: WIDGET_NAME,
        version: VERSION,
        description: 'UTC clock and uptime rendered from the xtop state (Node.js)',
        author: 'xtop examples',
        max_processes: 1, // this widget does not read processes
        api: '1',
      },
    });
  } else if (kind === 'render') {
    let ops = [];
    try {
      ops = buildOps(request.state);
    } catch (err) {
      sendLog('render failed: ' + err.message);
      ops = [];
    }
    send({ type: 'draw', ops });
  } else if (kind === 'shutdown') {
    return false;
  } else {
    sendLog('unknown request type: ' + kind);
  }
  return true;
}

const rl = readline.createInterface({ input: process.stdin, terminal: false });

rl.on('line', (line) => {
  if (!line.trim()) return; // ignore blank lines
  let request;
  try {
    request = JSON.parse(line);
  } catch (err) {
    sendLog('invalid JSON request: ' + err.message);
    return;
  }
  if (request === null || typeof request !== 'object' || Array.isArray(request)) {
    sendLog('request is not a JSON object');
    return;
  }
  if (!handle(request)) rl.close();
});

// Fires on shutdown and on stdin EOF; both exit 0.
rl.on('close', () => process.exit(0));
